docs/extend/frontend-pages.md

```diff
@@ -55,13 +55,95 @@ app.current.matches(DiscussionPage);
 
-Due to Mithril's routing implementation, `oninit` is NOT called again if a route is changed BUT the same component handles the new route.
-For instance, if you go directly from one discussion's page to another, `oninit` will not be called again, and the page will not be rerendered from scratch.
-See https://mithril.js.org/route.html#key-parameter. In core, we have identified 2 strategies that, when used together, replicate previous behavior.
+[Advanced use cases](https://mithril.js.org/route.html#advanced-component-resolution) can take advantage of Mithril's [route resolver system](https://mithril.js.org/route.html#routeresolver).
+Flarum actually already wraps all its components in the `flarum/resolvers/DefaultResolver` resolver. This has the following benefits:
 
-- If the route is programatically changed, and we always want to recreate the page component, we can use the `common/utils/setRouteWithForcedRefresh` util. Under the surface, this uses a unique key as per the above Mithril documentation.
-- When creating a page, we can store the current path (or some other identifying parameter) on initialization. Then, in `onbeforeupdate`, if the current value of that identifying parameter isn't the same as what we stored, we can load in new data for this page and re-render. In core, this is done in `IndexPage`, `DiscussionPage`, and `UserPage`. We recommend reading that source code as inspiration IF this is needed.
+- It passes a `routeName` attr to the current page, which then provides it to `PageState`
+- It assigns a [key](https://mithril.js.org/keys.html#single-child-keyed-fragments) to the top level page component. When the route changes, if the top level component's key has changed, it will be completely rerendered (by default, Mithril does not rerender components when switching from one page to another if both are handled by the same component).
 
-We are planning to provide a cleaner abstraction in future beta releases. In the meantime, please note that this is only necessary for page components that handle multiple routes.
+### Using Route Resolvers
+
+There are actually 3 ways to set the component / route resolver when registering a route:
+
+- the `resolver` key can be used to provide an **instance** of a route resolver. This instance should define which component should be used, and hardcode the route name to be passed into it. This instance will be used without any modifications by Flarum.
+- The `resolverClass` key AND `component` key can be used to provide a **class** that will be used to instantiate a route resolver, to be used instead of Flarum's default one, as well as the component to use. Its constructor should take 2 arguments: `(component, routeName)`.
+- The `component` key can be used alone to provide a component. This will result in the default behavior.
+
+For example:
+
+```js
+import CustomPage from './components/CustomPage';
+import CustomPageResolver from './resolvers/CustomPageResolver';
```

Comment:
I find it a bit confusing to show import statements, but not what the code for those would like.

I'd show maybe a shortened example, like

```js
class CustomPageResolver extends DefaultResolver {
    // See below for a full example of a custom resolver
}
```

Reply:
Modified it slightly, how's it now?

Reply:
Yeah, we can go with that :+1: